### PR TITLE
Painless: Add Bindings

### PR DESCRIPTION
**Pull request by jdconrad**
This change allows the specification of a data structure called a binding to store some permanent state within a single script instance.  An example binding specified in the whitelist would be as follows:

```
binding org.elasticsearch.painless.TestBinding {
    (int, int)
    int someCall(int, int)
}
```

Within the script a user would use this class by simply doing the following:

```someCall(1, 2, 3, 4)```

where the arguments are split between a call to the constructor the first time this method is called and then the method call for each time the method is called.  The constructor has the opportunity to store the arguments passed in as state within the TestBinding class for use within the member method.  Each subsequent time the script is run, the parameters are already stored in a local instance so they will not need to be re-calculated.

Each binding must contain exactly one constructor and one method.

### CODE REVIEW

**Comment by jdconrad** [comment]
After some quick discussion with @rjernst I will be changing the formatting of the whitelist to describe bindings shortly with a syntax we think is better than what's being proposed here.

**Comment by jdconrad** [comment]
@rjernst Updated formatting for the whitelist to the following as we spoke about privately.

```
static {
  int someMethod(int, int, int) bound_to org.package.someClass
}
```

File: modules/lang-painless/spi/src/main/java/org/elasticsearch/painless/spi/WhitelistLoader.java [hunk 1]
```diff
@@ -178,41 +184,125 @@ public static Whitelist loadFromResourceFiles(Class<?> resource, String... filep
...
 
                         // Reset all the constructors, methods, and fields to support a new class.
                         whitelistConstructors = new ArrayList<>();
                         whitelistMethods = new ArrayList<>();
                         whitelistFields = new ArrayList<>();
+                    } else if (line.startsWith("static ")) {
+                        // Ensure the final token of the line is '{'.
+                        if (line.endsWith("{") == false) {
+                            throw new IllegalArgumentException(
+                                    "invalid class definition: failed to parse class opening bracket [" + line + "]");
```
Comment (rjernst) [inline modules/lang-painless/spi/src/main/java/org/elasticsearch/painless/spi/WhitelistLoader.java]: Should the message here indicate "static" opening bracket, not class?

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/lookup/PainlessLookupBuilder.java [hunk 2]
```diff
@@ -731,6 +788,183 @@ public void addPainlessField(Class<?> targetClass, String fieldName, Class<?> ty
...
+
+        if (returnType == null) {
+            throw new IllegalArgumentException("return type [" + returnCanonicalTypeName + "] not found for binding " +
+                    "[[" + targetCanonicalClassName + "], [" + methodName + "], " + canonicalTypeNameParameters + "]");
+        }
+
+        addPainlessBinding(targetClass, methodName, returnType, typeParameters);
+    }
+
+    public void addPainlessBinding(Class<?> targetClass, String methodName, Class<?> returnType, List<Class<?>> typeParameters) {
```
Comment (rjernst) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/lookup/PainlessLookupBuilder.java]: Why two public variants of addPainlessBinding?

**Review by rjernst — APPROVED** [review]
LGTM, just a couple nits

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/lookup/PainlessLookupBuilder.java (hunk 2 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/lookup/PainlessLookupBuilder.java]: Consistency in the API.  All of the other add<Something> methods allow for things to be added directly as well without using Strings.  This allows flexibility for things like testing and such in the future where we don't necessarily want to be forced to use a whitelist.

File: modules/lang-painless/spi/src/main/java/org/elasticsearch/painless/spi/WhitelistLoader.java (hunk 1 above)
Comment (jdconrad) [reply modules/lang-painless/spi/src/main/java/org/elasticsearch/painless/spi/WhitelistLoader.java]: Fixed.

File: modules/lang-painless/spi/src/main/java/org/elasticsearch/painless/spi/WhitelistBinding.java [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+ * under the License.
+ */
+
+package org.elasticsearch.painless.spi;
+
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * A binding represents a static method call that stores state.  Each binding must have exactly
```
[the remark this answers is not in the document]
Comment (jdconrad) [reply modules/lang-painless/spi/src/main/java/org/elasticsearch/painless/spi/WhitelistBinding.java]: Cleaned this up a bit.

**Comment by jdconrad** [comment]
@rjernst Thanks for the review.  Will commit as soon as CI passes.

**Comment by jdconrad** [comment]
@mayya-sharipova This is ready to go.  Please let me know if you have any issues/need anything else for function scores.

**Comment by mayya-sharipova** [comment]
@jdconrad  Thanks a lot, Jack. Nice to see this change merged, will definitely try to use in function scores.

**Pull request merged**